File: JS/560-subarray-sum-equals-k.js
// prefix sum

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function (nums, k) {
  let map = new Map();
  let sum = 0,
    res = 0;
  map.set(0, 1);
  for (let num of nums) {
    sum += num;
    res += map.get(sum - k) || 0;
    map.set(sum, map.get(sum) + 1 || 1);
  }
  return res;
};

File: C++/560-subarray-sum-equals-k.cpp
#include "header.h"

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int res = 0, sum = 0;
        unordered_map<int, int> map;
        map[sum]++;
        for (int num : nums) {
            sum += num;
            res += map[sum - k];
            map[sum]++;
        }
        return res;
    }
};